toppra/dracula/__init__.py

```diff
@@ -1,8 +1,170 @@
+    # toppra only produces clamped cubic splines
+    path = ta.SplineInterpolator(x, waypts.copy(), bc_type="clamped")
+    pc_vel = constraint.JointVelocityConstraint(vlim)
+    # Can be either Collocation (0) or Interpolation (1). Interpolation gives
+    # more accurate results with slightly higher computational cost
+    pc_acc = constraint.JointAccelerationConstraint(
+        alim, discretization_scheme=constraint.DiscretizationType.Interpolation
+    )
+    # use the default gridpoints=None to let interpolator.propose_gridpoints
+    # calculate grid that sufficiently covers the path.
+    # this ensures the instance is controllable and avoids error:
+    #     "An error occurred when computing controllable velocities.
+    #     The path is not controllable, or is badly conditioned.
+    #     Error: Instance is not controllable"
+    # if using clamped as boundary condition, the default gridpoints error
+    # 1e-3 is ok and we don't need to calculate gridpoints
+    # boundary condition "natural" especially needs support by smaller error
+    # gridpoints = interpolator.propose_gridpoints(
+    #     path, max_err_threshold=max_grid_err
+    # )
+    instance = algo.TOPPRA(
+        [pc_vel, pc_acc],
+        path,
+        # gridpoints=gridpoints,
+        solver_wrapper="seidel",
+    )
+    jnt_traj = instance.compute_trajectory(0, 0)
+    if jnt_traj is None:  # toppra has failed
+        if min_pair_dist < JOINT_DIST_EPS:  # duplicates are probably why
+            print(
+                "Duplicates not allowed in input waypoints. "
+                "At least one pair of adjacent waypoints have "
+                f"joint-space distance less than epsilon = {JOINT_DIST_EPS}.",
+                file=sys.stderr,
+            )
+        print(
+            f"Failed waypts:\n{waypts}\n" f"vlim:\n{vlim}\n" f"alim:\n{alim}",
+            file=sys.stderr,
+        )
+        raise RuntimeError("Toppra failed to compute trajectory.")
+    cspl = jnt_traj.cspl
+    # If the initial estimated path length (run time) of the trajectory isn't
+    # very close to the actual computed one (say off by a factor of 2),
+    # toppra goes a bit crazy sometimes extends the spline to x >~ 1e3.
+    # This cannot be fixed by tweaking the fit:
+    # N_samples = 39 may have this behaviour unless x_max > 1.6,
+    # while N_samples = 46 may require x_max < 1.6 to be controllable.
+    # So we keep x_max towards the small side to guarantee controllability.
+    # We check that and truncate the spline at a predefined upper bound.
+    if path_length_limit and cspl.x[-1] > path_length_limit:
+        mask = cspl.x <= path_length_limit
+        warnings.warn(
+            f"Toppra derived x > {path_length_limit} covering "
+            f"{(~mask).sum()}/{cspl.x.size} ending knots, "
+            f"x_max: {cspl.x[-1]:.2f}. Input waypoints are likely ill-formed, "
+            "resulting in a lengthy trajectory.",
+            UserWarning,
+        )
```

Comment:
Should we promote this to an exception? This seems like the sort of thing that we may want to stop interpolation on

Reply:
no, because when this happens, often times it is "salvageable". namely there's only one or a few abnormal points outside the limit and everything else prior to them is fine, and truncating the points outside the limit doesn't change the ending position. it's an artefact that's easily identifiable and fixable.